File: python/problems/leetcode/0034_find_first_and_last_position_of_element_in_sorted_array/solution.py
"""LeetCode 34: Find First and Last Position of Element in Sorted Array."""

from __future__ import annotations

from bisect import bisect_left, bisect_right


class Solution:
    def searchRange(self, nums: list[int], target: int) -> list[int]:
        l = bisect_left(nums, target)
        r = bisect_right(nums, target) - 1
        if l <= r:
            return [l, r]
        return [-1, -1]


if __name__ == "__main__":
    assert Solution().searchRange([5, 7, 7, 8, 8, 10], 8) == [3, 4]
    assert Solution().searchRange([5, 7, 7, 8, 8, 10], 6) == [-1, -1]
    assert Solution().searchRange([], 0) == [-1, -1]
    print("0034 OK")

File: cpp/problems/leetcode/0034_find_first_and_last_position_of_element_in_sorted_array/solution.cpp
// LeetCode 34
#include <bits/stdc++.h>
#include <cassert>
using namespace std;

class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        auto lo = lower_bound(nums.begin(), nums.end(), target);
        auto hi = upper_bound(nums.begin(), nums.end(), target);
        if (lo == hi) return {-1, -1};
        return {(int)(lo - nums.begin()), (int)(hi - nums.begin() - 1)};
    }
};

int main() {
    vector<int> a{5, 7, 7, 8, 8, 10};
    assert((Solution().searchRange(a, 8) == vector<int>{3, 4}));
    assert((Solution().searchRange(a, 6) == vector<int>{-1, -1}));
    vector<int> e{};
    assert((Solution().searchRange(e, 0) == vector<int>{-1, -1}));
    cout << "0034 OK" << endl;
    return 0;
}
